### ISSUE TITLE
Invalid constructor SFINAE for std::shared_ptr's array ctors

### ISSUE THREAD

**Issue by ldionne** [llvm/llvm-project#60258]
This was reported to me by Marshall Clow (thanks!)

The SFINAE criteria for `shared_ptr`'s constructors don't seem right. Consider the following code:

```c++
#include <tuple>
#include <cassert>
#include <memory>

int main () {
	std::shared_ptr<int> i;
	std::shared_ptr<int[5]> i5a = i;
	std::shared_ptr<int[]> iUa = i;
	std::shared_ptr<int[]> iUb = i5a;
	std::shared_ptr<int[5]> i5b = iUa;
	std::shared_ptr<int[7]> i7 = i5a;
}
```

Clang + libc++ accepts it (C++17).
GCC + libstdc++ rejects it (C++17)
MSVC rejects it (C++17)

```console
<source>(7): error C2440: 'initializing': cannot convert from 'std::shared_ptr<int>' to 'std::shared_ptr<int [5]>'
<source>(8): error C2440: 'initializing': cannot convert from 'std::shared_ptr<int>' to 'std::shared_ptr<int []>'
<source>(10): error C2440: 'initializing': cannot convert from 'std::shared_ptr<int []>' to 'std::shared_ptr<int [5]>'
<source>(11): error C2440: 'initializing': cannot convert from 'std::shared_ptr<int [5]>' to 'std::shared_ptr<int [7]>'
```

**Comment by ldionne** [comment]
@huixie90 Could you take a look? I think you just tinkered with `std::shared_ptr`'s `make_shared` functions, so perhaps this is fresh in your memory?

**Comment by huixie90** [comment]
> @huixie90 Could you take a look? I think you just tinkered with `std::shared_ptr`'s `make_shared` functions, so perhaps this is fresh in your memory?

sure

**Comment by huixie90** [comment]
ATM, we unconditionally `remove_extent_t` whiling checking the constraints
```
template<class _Tp, class _Up>
struct __compatible_with
#if _LIBCPP_STD_VER > 14
    : is_convertible<remove_extent_t<_Tp>*, remove_extent_t<_Up>*> {};
#else
    : is_convertible<_Tp*, _Up*> {};
#endif // _LIBCPP_STD_VER > 14
```

https://github.com/llvm/llvm-project/blob/main/libcxx/include/__memory/shared_ptr.h#L373

**Comment by huixie90** [comment]
This should be fixed by
https://github.com/llvm/llvm-project/commit/a38a4654ce4b1d2ae8a03797d2e520e415150492
